compiler/rustc_mir/src/transform/unneeded_deref.rs

```diff
@@ -129,7 +131,7 @@ impl<'a, 'tcx> ResultsVisitor<'a, 'tcx> for UnneededDerefVisitor<'a, 'tcx> {
         stmt: &'mir Statement<'tcx>,
         location: Location,
     ) {
-        self.state = Some(state.clone());
+        self.state = state;
```

Comment:
There's no need to use `self` for the `visit_rvalue`. You could also create a separate visitor that has a reference to `Self` as a field and thus can have a reference with a lifetime in the `state` field. If performance is the same, then I would definitely prefer that.